File: lib/models/espece.dart
class Espece {
  final String id;
  final String nom;
  final String description;
  final String? photo;
  final double prixMinimal;
  final double prixMoyen;
  final bool isActive;

  Espece({
    required this.id,
    required this.nom,
    this.description = '',
    this.photo,
    this.prixMinimal = 0.0,
    this.prixMoyen = 0.0,
    this.isActive = true,
  });

  factory Espece.fromMap(Map<String, dynamic> map) {
    return Espece(
      id: map['id']?.toString() ?? map['_id']?.toString() ?? '',
      nom: map['nom'] ?? '',
      description: map['description'] ?? '',
      photo: map['photo'] ?? map['image_url'],
      prixMinimal:
          _parseDouble(map['prixMinimal'] ?? map['prix_minimal']) ?? 0.0,
      prixMoyen: _parseDouble(map['prixMoyen'] ?? map['prix_moyen']) ?? 0.0,
      isActive: map['isActive'] ?? map['is_active'] ?? true,
    );
  }

  Map<String, dynamic> toMap() {
    return {
      'id': id,
      'nom': nom,
      'description': description,
      'photo': photo,
      'prixMinimal': prixMinimal,
      'prixMoyen': prixMoyen,
      'isActive': isActive,
    };
  }

  @override
  String toString() {
    return 'Espece{id: $id, nom: $nom, description: $description, photo: $photo, prixMinimal: $prixMinimal, prixMoyen: $prixMoyen, isActive: $isActive}';
  }

  /// Convertit une valeur en double
  static double? _parseDouble(dynamic value) {
    if (value == null) return null;
    if (value is double) return value;
    if (value is int) return value.toDouble();
    if (value is String) {
      try {
        return double.parse(value);
      } catch (_) {
        return null;
      }
    }
    return null;
  }
}

File: backend/src/models/Espece.js
const mongoose = require('mongoose');

const especeSchema = new mongoose.Schema({
  nom: {
    type: String,
    required: true,
    trim: true
  },
  imageUrl: {
    type: String,
    required: true
  },
  description: {
    type: String,
    trim: true
  },
  nomScientifique: {
    type: String,
    trim: true
  },
  prixMinimal: {
    type: Number,
    default: 0
  },
  prixMoyen: {
    type: Number,
    default: 0
  },
  isActive: {
    type: Boolean,
    default: true
  },
  // Nouvelles propriétés pour la classification des poissons
  confiance: {
    type: Number,
    default: 1.0
  },
  source: {
    type: String,
    enum: ['manuel', 'api', 'tensorflow', 'google_vision'],
    default: 'manuel'
  },
  alternatives: [{
    nom: String,
    nomScientifique: String,
    confiance: Number
  }],
  // Propriétés pour la traçabilité
  saison: {
    type: String,
    trim: true
  },
  habitat: {
    type: String,
    trim: true
  },
  methodePeche: {
    type: String,
    trim: true
  }
}, {
  timestamps: true
});

// Ajouter un index texte pour améliorer les performances de recherche
especeSchema.index({ nom: 'text', nomScientifique: 'text' });

// Ajouter des index pour les champs fréquemment utilisés dans les requêtes
especeSchema.index({ nom: 1 });
especeSchema.index({ isActive: 1 });
especeSchema.index({ prixMinimal: 1 });
especeSchema.index({ prixMoyen: 1 });

const Espece = mongoose.model('Espece', especeSchema);

module.exports = Espece;